src/base/bittorrent/session.cpp

```diff
@@ -1065,6 +1083,266 @@ void Session::configureComponents()
 #endif
 }
 
+void Session::prepareStartup()
+{
+    qDebug("Initializing torrents resume data storage...");
+
+    const Path dbPath = specialFolderLocation(SpecialFolder::Data) / Path(u"torrents.db"_qs);
+    const bool dbStorageExists = dbPath.exists();
+
+    ResumeDataStorage *startupStorage = nullptr;
+    if (resumeDataStorageType() == ResumeDataStorageType::SQLite)
+    {
+        m_resumeDataStorage = new DBResumeDataStorage(dbPath, this);
+
+        if (!dbStorageExists)
+        {
+            const Path dataPath = specialFolderLocation(SpecialFolder::Data) / Path(u"BT_backup"_qs);
+            startupStorage = new BencodeResumeDataStorage(dataPath, this);
+        }
+    }
+    else
+    {
+        const Path dataPath = specialFolderLocation(SpecialFolder::Data) / Path(u"BT_backup"_qs);
+        m_resumeDataStorage = new BencodeResumeDataStorage(dataPath, this);
+
+        if (dbStorageExists)
+            startupStorage = new DBResumeDataStorage(dbPath, this);
+    }
+
+    if (!startupStorage)
+        startupStorage = m_resumeDataStorage;
+
+    auto *context = new ResumeSessionContext(this);
+    context->startupStorage = startupStorage;
+
+    connect(startupStorage, &ResumeDataStorage::loadStarted, context
+            , [this, context](const QVector<TorrentID> &torrents)
+    {
+#ifdef QBT_USES_LIBTORRENT2
+        context->indexedTorrents = QSet<TorrentID>(torrents.cbegin(), torrents.cend());
+#endif
+
+        handleLoadedResumeData(context);
+    });
+
+    connect(startupStorage, &ResumeDataStorage::loadFinished, context, [context]()
+    {
+        context->isLoadFinished = true;
+    });
+
+    connect(this, &Session::torrentsLoaded, context, [this, context](const QVector<Torrent *> &torrents)
+    {
+        context->processingResumeDataCount -= torrents.count();
+        QMetaObject::invokeMethod(this, [this, context]() { handleLoadedResumeData(context); }, Qt::QueuedConnection);
+        m_nativeSession->post_torrent_updates();
+        m_refreshEnqueued = true;
+    });
+
+    startupStorage->loadAll();
+}
+
+void Session::handleLoadedResumeData(ResumeSessionContext *context)
+{
+    while (context->processingResumeDataCount < MAX_PROCESSING_RESUMEDATA_COUNT)
+    {
+        if (context->loadedResumeData.isEmpty())
+            context->loadedResumeData = context->startupStorage->fetchLoadedResumeData();
+
+        if (context->loadedResumeData.isEmpty())
+        {
+            if (context->processingResumeDataCount == 0)
+            {
+                if (context->isLoadFinished)
+                    endStartup(context);
+                else
+                    QMetaObject::invokeMethod(this, [this, context]() { handleLoadedResumeData(context); }, Qt::QueuedConnection);
```

Comment:
is this rescheduling itself? won't it keep reschedule forever?

Reply:
> is this rescheduling itself? won't it keep reschedule forever?

It is inside conditional statement.

Reply:
Well, redundant scheduling still turned out to be possible under certain conditions (not the ones I originally thought of). So now I'm really tracking whether it was already scheduled.